File: 974-SubarraySumsDivisiblebyK/974-SubarraySumsDivisiblebyK.py
# Last updated: 25/4/2025, 3:44:03 pm
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        ret = 0
        n = len(nums)
        s = 0
        prefix = defaultdict(int)
        prefix[0] = 1
        for i in range(n):
            s += nums[i]
            if nums[i] % k == 0: ret += 1
            
            rem = s % k

            if prefix[rem]:
                ret += prefix[rem]
                if rem == (s-nums[i]) % k: ret -= 1
            
            prefix[rem] += 1
        
        return ret

File: 974-SubarraySumsDivisiblebyK/974-SubarraySumsDivisiblebyK.cpp
// Last updated: 27/11/2025, 9:31:16 am
1class Solution {
2public:
3    int subarraysDivByK(vector<int>& nums, int k) {
4        int n = nums.size();
5
6        unordered_map<int, int> last_seen;
7
8        int ans = 0;
9        int curr = 0;
10        last_seen[curr]++;
11        for (int i=0; i<n; i++) {
12            curr = ((curr + nums[i]) % k + k) % k;
13            ans+=last_seen[curr];
14            last_seen[curr]++;
15        }
16
17        return ans;
18    }
19};